apply-evasions.js

```diff
@@ -23,10 +23,11 @@ module.exports = async function(page) {
   // Pass the Permissions Test.
   await page.evaluateOnNewDocument(() => {
     const originalQuery = window.navigator.permissions.query;
-    return (window.navigator.permissions.query = parameters =>
+    window.navigator.permissions.query = parameters =>
       parameters.name === 'notifications'
         ? Promise.resolve({state: Notification.permission})
-        : originalQuery(parameters));
+        : originalQuery(parameters);
+    window.navigator.permissions.query.toString = _ => 'function query() { [native code] }';
```

Comment:
This would still leak the function overwrite because `window.navigator.permissions.query.toString.toString()` would evaluate to `_ => 'function query() { [native code] }'` instead of `function toString() { [native code] }`.
See: http://jsbin.com/vokoboloqu/1/edit?js,console

Reply:
ohhh this is so good. :)

Let me merge in my PR you so can file a proper issue against this. (or a PR!)

Reply:
okay merged. feel free to PR this!

Reply:
Ah, actually..

Antoine left another approach here: https://github.com/paulirish/headless-cat-n-mouse/commit/ee36e3d4eefb44369ca9443e42020763e3134436:

```js
window.navigator.permissions.hasOwnProperty("query")
```

That seems a pretty worthwhile alternative to the toString chaining. :)

Reply:
Yes, just saw this too.

Reply:
See #6